cartography/intel/aws/ec2/snapshots.py

```diff
@@ -0,0 +1,110 @@
+import logging
+from typing import Dict
+from typing import List
+
+import boto3
+import neo4j
+
+from cartography.util import aws_handle_regions
+from cartography.util import run_cleanup_job
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+
+
+@timeit
+@aws_handle_regions
+def get_snapshots(boto3_session: boto3.session.Session, region: str) -> List[Dict]:
+    client = boto3_session.client('ec2', region_name=region)
+    paginator = client.get_paginator('describe_snapshots')
+    snapshots: List[Dict] = []
+    for page in paginator.paginate():
+        snapshots.extend(page['Snapshots'])
+    return snapshots
+
+
+@timeit
+def load_snapshots(
+        neo4j_session: neo4j.Session, data: List[Dict], region: str, current_aws_account_id: str, update_tag: int,
+) -> None:
+    ingest_snapshots = """
+    UNWIND {snapshots_list} as snapshot
+    MERGE (s:EBSSnapshot{id: snapshot.SnapshotId})
+    ON CREATE SET s.firstseen = timestamp()
+    SET s.lastupdated = {update_tag}, s.description = snapshot.Description, s.encrypted = snapshot.Encrypted,
+    s.progress = snapshot.Progress, s.starttime = snapshot.StartTime, s.state = snapshot.State,
+    s.statemessage = snapshot.StateMessage, s.volumeid = snapshot.VolumeId, s.volumesize = snapshot.VolumeSize,
+    s.outpostarn = snapshot.OutpostArn, s.dataencryptionkeyid = snapshot.DataEncryptionKeyId,
+    s.kmskeyid = snapshot.KmsKeyId, s.region={Region}
+    WITH s
+    MATCH (aa:AWSAccount{id: {AWS_ACCOUNT_ID}})
+    MERGE (aa)-[r:RESOURCE]->(s)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {update_tag}
+    """
+
+    # neo4j does not accept datetime objects and values. This loop is used to convert
+    # these values to string.
+    for snapshot in data:
+        snapshot['StartTime'] = str(snapshot['StartTime'])
+
+    neo4j_session.run(
+        ingest_snapshots,
+        snapshots_list=data,
+        AWS_ACCOUNT_ID=current_aws_account_id,
+        Region=region,
+        update_tag=update_tag,
+    )
+
+
+@timeit
+def load_snapshot_volume_relations(
+        neo4j_session: neo4j.Session, data: List[Dict], current_aws_account_id: str, update_tag: int,
+) -> None:
+    ingest_volumes = """
+    MERGE (v:EBSVolume{id: {VolumeId})
+    ON CREATE SET v.firstseen = timestamp()
+    SET v.lastupdated = {update_tag}
+    WITH v
+    MATCH (aa:AWSAccount{id: {AWS_ACCOUNT_ID}})
+    MERGE (aa)-[r:RESOURCE]->(v)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {update_tag}
+    WITH v
+    MATCH (s:EBSSnapshot{id: {SnapshotId}})
+    MERGE (s)-[r:CREATED_FROM]->(v)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {update_tag}
+    """
+
+    for snapshot in data:
+        if snapshot.get('VolumeId'):
+            neo4j_session.run(
```

Comment:
We should try to use unwind here, instead of loop based insert

Reply:
We need to check if VolumeId is present or not in the snapshot - only then should we ingest otherwise it creates nodes with null ids and the cleanup code errors out. Hence I stuck to the loop method here. Let's keep this for now, I'll look at an optimum way for it next week  and open a PR for that if needed.

Reply:
One way is to create a new dictionary and add snapshot data in it if VolumeId is present. Should I do that then?

Reply:
yes, we should do that. so that we are loading only snapshots which has `VolumeId`

Reply:
Ok!